UserSessionsMeter.c

```diff
@@ -0,0 +1,157 @@
+#include "generic/utmpx.h"
+
+#if defined(BUILD_STATIC) && defined(HAVE_LIBSYSTEMD)
+#include <systemd/sd-login.h>
+#endif
+
+#ifndef BUILD_STATIC
+#include <dlfcn.h>
+#include "linux/Platform.h"
+#endif
+
+
+static int users = 0;
+
+#ifdef BUILD_STATIC
+#define sym_sd_get_sessions sd_get_sessions
+#else
+
+# ifdef HTOP_LINUX
+static int (*sym_sd_get_sessions)(char***);
+static void* dlopenHandle;
+static size_t activeHandles;
+# endif /* HTOP_LINUX */
+
+#endif /* BUILD_STATIC */
+
+
+static void UserSessionsMeter_init(ATTR_UNUSED Meter* this) {
+#if !defined(BUILD_STATIC) && defined(HTOP_LINUX)
+   assert(activeHandles != SIZE_MAX);
+   assert((dlopenHandle != NULL) == (activeHandles > 0));
+
+   if (activeHandles > 0) {
+      activeHandles++;
+      return;
+   }
+
+   dlopenHandle = Platform_load_libsystemd();
+   if (!dlopenHandle)
+      return;
+
+   /* Clear any errors */
+   dlerror();
+
+   #define resolve(symbolname) do {                                      \
+      *(void **)(&sym_##symbolname) = dlsym(dlopenHandle, #symbolname);  \
+      if (!sym_##symbolname || dlerror() != NULL)                        \
+         goto dlfailure;                                                 \
+   } while(0)
+
+   resolve(sd_get_sessions);
+
+   #undef resolve
+
+   activeHandles++;
+   return;
+
+dlfailure:
+   Platform_close_libsystemd();
+   dlopenHandle = NULL;
+#endif /* !BUILD_STATIC && HTOP_LINUX */
+}
+
+static void UserSessionsMeter_done(ATTR_UNUSED Meter* this) {
+#if !defined(BUILD_STATIC) && defined(HTOP_LINUX)
+   if (activeHandles > 0) {
+      activeHandles--;
+      if (activeHandles == 0) {
+         Platform_close_libsystemd();
+         dlopenHandle = NULL;
+      }
+   }
+#endif /* !BUILD_STATIC && HTOP_LINUX */
+}
+
+#if (!defined(BUILD_STATIC) && defined(HTOP_LINUX)) || defined(HAVE_LIBSYSTEMD)
+static int update_via_sd_login(void) {
+#ifndef BUILD_STATIC
+   if (!dlopenHandle)
+      return -1;
+#endif /* !BUILD_STATIC */
+
+   return sym_sd_get_sessions(NULL);
+}
+#endif /* (!BUILD_STATIC && HTOP_LINUX) || HAVE_LIBSYSTEMD */
+
+static void UserSessionsMeter_updateValues(Meter* this) {
+   int ret;
+#if (!defined(BUILD_STATIC) && defined(HTOP_LINUX)) || defined(HAVE_LIBSYSTEMD)
+   ret = update_via_sd_login();
+   if (ret < 0)
+      ret = Generic_utmpx_get_users();
+#else
+   ret = Generic_utmpx_get_users();
+#endif /* (!BUILD_STATIC && HTOP_LINUX) || HAVE_LIBSYSTEMD */
+
+   users = ret;
+
+   if (users >= 0) {
+      xSnprintf(this->txtBuffer, sizeof(this->txtBuffer), "%d", ret);
+   } else {
+      xSnprintf(this->txtBuffer, sizeof(this->txtBuffer), "N/A");
+   }
+}
+
+static void UserSessionsMeter_display(ATTR_UNUSED const Object* cast, RichString* out) {
+   char buffer[16];
+
+   if (users >= 0) {
+      int len = xSnprintf(buffer, sizeof(buffer), "%d", users);
+      RichString_appendnAscii(out, CRT_colors[METER_VALUE], buffer, len);
+   } else {
+      RichString_appendAscii(out, CRT_colors[METER_VALUE], "N/A");
```

Comment:
Maybe shadow the value if not available?

Reply:
For meter text, I don't think shadowing is a good idea. IMHO.